[Cookie Store API] Ensure secure property defaults to true

RupinMittal opened the pull request:
#### 44f403689534bcd565dfeebea93c1dd68f63a9ab
<pre>
[Cookie Store API] Ensure secure property defaults to true
<a href="https://bugs.webkit.org/show_bug.cgi?id=285623">https://bugs.webkit.org/show_bug.cgi?id=285623</a>
<a href="https://rdar.apple.com/142442877">rdar://142442877</a>

Reviewed by Chris Dumez and Sihui Liu.

The Cookie Store API spec (<a href="https://wicg.github.io/cookie-store/#secure-cookies)">https://wicg.github.io/cookie-store/#secure-cookies)</a>
says this API should only set cookies as secure and when fetching/modifying non-secure
cookies, it should modify them to be secure as well.

With this patch:
1. CookieStore::set() will now set the secure propety to true when setting a cookie.
2. CookieListItem (which is what is constructed from a cookie and returned when this API
   fetches cookies) will always have secure as true.
3. As a consequence of (2), if CookieStore::get() or getAll() are used to fetch an un-secure
   cookie, they will set it to secure.

Testing:
1. New layout test confirms that CookieListItem defaults to true (will be upstreamed to WPT)
2. New layout test confirms that CookieStore::set always sets a secure cookie
3. New layout test confirms that if CookieStore::get is used to get a un-secure cookie, it
   will set that cookie to be secure.

* LayoutTests/http/tests/cookies/cookie-store-get-secure.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/cookie-store-get-secure.https.html: Added.
* LayoutTests/http/tests/cookies/cookie-store-set-secure.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/cookie-store-set-secure.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.js:
(async set promise_test):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookies/partitioned-cookies/partitioned-cookies.tentative.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setCookie):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/288824@main">https://commits.webkit.org/288824@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5dc632f910796857d834efa461d7673a35f0a37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4132 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38812 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89589 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86592 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4217 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65719 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23562 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87553 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3188 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76775 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46013 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3136 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31006 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34566 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31781 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90970 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11774 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74171 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12001 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72599 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73372 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17714 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16157 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3209 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13162 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11726 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17202 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11575 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15051 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13348 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.js:
```diff
@@ -195,3 +195,15 @@ promise_test(async testCase => {
+  });
+
+  const cookie = await cookieStore.get('cookie-name');
+  assert_equals(cookie.secure, true);
```
We might also want to add a test to ensure `document.cookie` agrees with result of `cookieStore.get()` on secure attribute

RupinMittal replied:
I might be mistaken, but I believe there's no way to check the secure attribute from document.cookie. Is there a way?

szewai replied:
There is no "secure" in the output, but we could check when cookie is secure indirectly by loading `document.cookie` in non-secure context (e.g. if you set the cookie with cookie store API in https://example.com, check if "document.cookie" in http://example.com contains the cookie), or you may check the HTTP request header, see examples in LayoutTests/http/tests/cookies/ -- mainly to ensure we actually store secure cookie.

The current implementation will always set `CookieListItem::secure` to true (`CookieListItem(Cookie&& cookie)`) regardless of the actual value, so the check (`assert_equals(cookie.secure, true);`) does not prove the cookie created by Cookie Store API is secure.

RupinMittal replied:
Re-uploaded with tests that check set(), get(), and getAll().

RupinMittal committed:
[Cookie Store API] Ensure secure property defaults to true
https://bugs.webkit.org/show_bug.cgi?id=285623
rdar://142442877

Reviewed by Chris Dumez and Sihui Liu.

The Cookie Store API spec (https://wicg.github.io/cookie-store/#secure-cookies)
says this API should only set cookies as secure and when fetching/modifying non-secure
cookies, it should modify them to be secure as well.

With this patch:
1. CookieStore::set() will now set the secure propety to true when setting a cookie.
2. CookieListItem (which is what is constructed from a cookie and returned when this API
   fetches cookies) will always have secure as true.
3. As a consequence of (2), if CookieStore::get() or getAll() are used to fetch an un-secure
   cookie, they will set it to secure.

Testing:
1. New layout test confirms that CookieListItem defaults to true (will be upstreamed to WPT)
2. New layout test confirms that CookieStore::set always sets a secure cookie
3. New layout test confirms that if CookieStore::get is used to get a un-secure cookie, it
   will set that cookie to be secure.

* LayoutTests/http/tests/cookies/cookie-store-get-secure.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/cookie-store-get-secure.https.html: Added.
* LayoutTests/http/tests/cookies/cookie-store-set-secure.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/cookie-store-set-secure.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.js:
(async set promise_test):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookies/partitioned-cookies/partitioned-cookies.tentative.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setCookie):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/288824@main
